Change RUN and CDUMP for ensemble jobs

WalterKolczynski-NOAA opened the pull request:
**Description**
Updates `$RUN` and `$CDUMP` for ensemble jobs to include `enkf`.

Previously, the `$RUN` for EnKF jobs had been set to `gdas` or `gfs`. However, this violates NCO policy and was also complicating Issue #761 . Now the `$RUN` for EnKF jobs is either `enkfgdas` or `enkfgfs`.

Theoretically, `$CDUMP` shouldn't need to change. However, `$CDUMP` and `$RUN` are used interchangeably throughout much of the workflow (Issue #1299), so for now the `$CDUMP` is kept identical to `$RUN`. This will be corrected in a future PR.

This change **changes the name** of enkf output files. Files now begin `enkfgdas` (or `enkfgfs`) rather than `gdas`.

Closes #1298

**Type of change**
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

**How Has This Been Tested?**
- [x] Cycled test on Orion
  
**Checklist**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes

WalterKolczynski-NOAA committed:
Change RUN for ensemble jobs

Updates `$RUN` and `$CDUMP` for ensemble jobs to include `enkf`.

Previously, the `$RUN` for EnKF jobs had been set to `gdas` or `gfs`.
However, this violates NCO policy and was also complicating Issue #761.
Now the `$RUN` for EnKF jobs is either `enkfgdas` or `enkfgfs`.

Theoeretically, `$CDUMP` shouldn't need to change. However, `$CDUMP`
and `$RUN` are used interchangably throughout much of the workflow
(Issue #1299), so for now the `$CDUMP` is kept identical to `$RUN`.

Refs: #1298

WalterKolczynski-NOAA committed:
Fix analysis location for eobs

The enkf obs job was looking in the wrong directory for the
deterministic analysis files following the prior commit changing
the `$RUN` for EnKF.

Refs: #1298

WalterKolczynski-NOAA commented:
Leaving draft for now until I do a bit more testing and figure out how we will address the GDAS App jobs. So far, I have not updated the GDAS App jobs, but they will be impacted by this change as well. Need to discuss how we want to handle those.

WalterKolczynski-NOAA committed:
Fix syntax error with pattern matching

Corrects a syntax error with bash pattern matching, which does
not work in single square braces.

Refs: #1298

WalterKolczynski-NOAA committed:
Fix linter complaint

WalterKolczynski-NOAA commented:
@CoryMartin-NOAA @guillaumevernieres @HenryWinterbottom-NOAA
This change is going to require updates to UFSDA as well. Also note a couple other expected changes along the same lines so you're aware and don't think we are just repeatedly changing the same stuff without a plan:

- This `$RUN` change will be permanent, but `$CDUMP` will be reverted back once workflow has been updated to use it properly instead of as a stand-in for `$RUN`
- More fundamental changes are coming via the use of path templates as part of the #761. You can get an **early** sneak preview here: https://github.com/WalterKolczynski-NOAA/global-workflow/blob/feature/com_reorg/parm/config/config.com

WalterKolczynski-NOAA commented:
Outside of the filename changes, output for this PR confirmed to be bitwise-identical after 4½ cycles. A prototype coupled was also tested and found to be identical.

KateFriedman-NOAA commented on jobs/JGDAS_ATMOS_CHGRES_FORENKF:
```diff
@@ -17,31 +17,19 @@ export DO_CALC_ANALYSIS=${DO_CALC_ANALYSIS:-"YES"}
-export COMIN_GES="${ROTDIR}/${GDUMP}.${gPDY}/${gcyc}/atmos"
-export COMIN_GES_ENS="${ROTDIR}/enkfgdas.${gPDY}/${gcyc}"
+
+export ATMFCST_ENSRES="${COMOUT_ENS}/mem001/atmos/${APREFIX_ENS}atmf006.nc"
```
Curious about this line. This seems even more hardcoded than similar lines like this in other scripts. Example from `jobs/JGDAS_ATMOS_ANALYSIS_DIAG`:
```
export ATMGES_ENSMEAN="${COMIN_GES_ENS}/${GPREFIX_ENS}atmf006.ensmean.nc"
```
Is it not possible to set `ATMFCST_ENSRES` without hardcoding `mem001/atmos`? Or do we need this hardcoding?

WalterKolczynski-NOAA replied:
I believe it is just needs an example of the file so it can extract the grid information. Since the number of members can vary, it is best to look for the first member, since that will be there no matter what the ensemble size.

KateFriedman-NOAA replied:
Ok, gotcha, thanks!

KateFriedman-NOAA commented on scripts/exgdas_enkf_earc.sh:
```diff
@@ -190,9 +190,9 @@ if [ "${GDATE}" -lt "${RDATE}" ]; then
 fi
 rPDY=$(echo "${RDATE}" | cut -c1-8)
-clist="gdas gfs"
+clist="gdas gfs enkfgdas enkfgfs"
```
Why still list `gdas gfs` in `clist` if this script is only for enkf archival?

WalterKolczynski-NOAA replied:
Good catch.

KateFriedman-NOAA reviewed (approved):
Couple small comments but approve!

WalterKolczynski-NOAA committed:
Remove earc cleanup of determinstic directories

The earc job does not need to clean up the deterministic ROTDIR
directories. The behavior was inadvertently changed during the update
for the enkf `$RUN`.

aerorahul commented on jobs/JGDAS_ENKF_DIAG:
```diff
@@ -22,31 +22,33 @@ GDATE=$(${NDATE} -${assim_freq} ${CDATE})
 gPDY=$(echo ${GDATE} | cut -c1-8)
 gcyc=$(echo ${GDATE} | cut -c9-10)
 GDUMP=${GDUMP:-"gdas"}
+GDUMP_ENS=${GDUMP_ENS:-"enkf${GDUMP}"}
```
This is a j-job. There is no one above this job setting `GDUMP_ENS`.
```suggestion
GDUMP_ENS="enkf${GDUMP}"
```
Same on the line above for `GDUMP`.

aerorahul commented on jobs/JGDAS_ENKF_DIAG:
```diff
@@ -22,31 +22,33 @@ GDATE=$(${NDATE} -${assim_freq} ${CDATE})
 gPDY=$(echo ${GDATE} | cut -c1-8)
 gcyc=$(echo ${GDATE} | cut -c9-10)
 GDUMP=${GDUMP:-"gdas"}
```
```suggestion
GDUMP="gdas"
```

WalterKolczynski-NOAA replied:
This wasn't something I changed, but I've changed it now to what you suggested (I had wondered the same thing).

aerorahul commented on jobs/JGDAS_ENKF_DIAG:
```diff
@@ -22,31 +22,33 @@ GDATE=$(${NDATE} -${assim_freq} ${CDATE})
 export CASE=${CASE_ENKF}
-export CDUMP_OBS=${CDUMP_OBS:-${CDUMP}}
+export CDUMP_OBS=${CDUMP_OBS:-${CDUMP/enkf}}
+export CDUMP_CTL=${CDUMP_CTL:-${CDUMP/enkf}}
```
What is intent of introducing `CDUMP_CTL`?

aerorahul replied:
Same for `CDUMP_OBS`

WalterKolczynski-NOAA replied:
`CDUMP_OBS` was a pre-existing variable. I created `CDUMP_CTL` so as not to misuse `CDUMP_OBS` for the deterministic member (even though they will be the same for now). I think this all goes away in the COM refactor when I revert `CDUMP` and use it per our discussion for both of these.

aerorahul commented on jobs/JGDAS_ENKF_ECEN:
```diff
@@ -20,35 +20,23 @@ GDATE=$(${NDATE} -${assim_freq} ${CDATE})
 gPDY=$(echo ${GDATE} | cut -c1-8)
 gcyc=$(echo ${GDATE} | cut -c9-10)
 GDUMP=${GDUMP:-"gdas"}
+GDUMP_ENS=${GDUMP_ENS:-"enkf${GDUMP}"}
```
Similar comment as the one made above.

aerorahul commented on jobs/JGDAS_ENKF_ECEN:
```diff
@@ -20,35 +20,23 @@ GDATE=$(${NDATE} -${assim_freq} ${CDATE})
 
 export CASE=${CASE_ENKF}
 
+export CDUMP_OBS=${CDUMP_OBS:-${CDUMP/enkf}}
```
`CDUMP_OBS` is needed for?

WalterKolczynski-NOAA replied:
This was done because this is how other j-jobs did it (though I clearly wasn't consistent, since I didn't use a `CDUMP_CTL` and instead just did the substitution in-line). I didn't worry _too_ much about consistency since I know this is all about to be substantially refactored anyway.

aerorahul reviewed (commented):
Can you explain the some basic choices on introducing variables that seem not necessary.
I will continue reviewing.

aerorahul commented:
@CoryMartin-NOAA @guillaumevernieres 
This change will impact GDASapp.
Please be aware of these inbound changes, and that they will change shortly as this is work in progress to resolve several outstanding EE2 issues.

CoryMartin-NOAA commented:
@aerorahul I need to discuss with you how to handle additional obs now that `COMIN_OBS` has been changed
